### PR TITLE
runtime-sdk: support dynamic min-gas-price

### PR DESCRIPTION
**Pull request by ptrus**
Adds support for dynamic min gas price inspired by EIP-1559

### CODE REVIEW

**Review by kostko — COMMENTED** [review]
I think this looks right, just some nits and comments around handling of edge cases (would be good to have tests for those as well).

File: runtime-sdk/src/modules/core/mod.rs [hunk 1]
```diff
@@ -23,15 +23,15 @@ use crate::{
...
+        token::{self, Denomination},
         transaction::{
             self, AddressSpec, AuthProof, Call, CallFormat, CallerAddress, UnverifiedTransaction,
         },
     },
     Runtime,
 };
 
-use self::types::RuntimeInfoResponse;
+use self::{state::DYNAMIC_MIN_GAS_PRICE, types::RuntimeInfoResponse};
```
[the remark this answers is not in the document]
Comment (ptrus) [reply runtime-sdk/src/modules/core/mod.rs]: :+1:

File: runtime-sdk/src/modules/core/mod.rs [hunk 2]
```diff
@@ -239,10 +268,25 @@ pub struct Parameters {
...
+    type Error = ParameterValidationError;
+
+    fn validate_basic(&self) -> Result<(), Self::Error> {
+        // Validate dynamic min gas price parameters.
+        let dmgp = &self.dynamic_min_gas_price;
+        if dmgp.enabled {
+            if dmgp.target_block_gas_usage_percentage > 100 {
+                return Err(ParameterValidationError::InvalidTargetBlockGasUsagePercentage);
+            }
+            if dmgp.min_price_max_change_denominator > 50 {
```
[the remark this answers is not in the document]
Comment (ptrus) [reply runtime-sdk/src/modules/core/mod.rs]: Indeed, updated.

File: runtime-sdk/src/modules/core/mod.rs [hunk 3]
```diff
@@ -1022,6 +1088,35 @@ impl<Cfg: Config> module::TransactionHandler for Module<Cfg> {
...
+///  - If the actual gas used is greater than the target gas used, increase the minimum gas price.
+///  - If the actual gas used is less than the target gas used, decrease the minimum gas price.
+///
+/// The price change is controlled by the `min_price_max_change_denominator` parameter.
+fn min_price_update(
+    gas_used: u128,
+    target_gas_used: u128,
+    min_price_max_change_denominator: u128,
+    current_price: u128,
+) -> u128 {
```
[the remark this answers is not in the document]
Comment (ptrus) [reply runtime-sdk/src/modules/core/mod.rs:1107]: Good idea, updated and also added test cases to `test_min_gas_price_update` for these "invalid" scenarios.

File: runtime-sdk/src/modules/core/mod.rs [hunk 4]
```diff
@@ -239,10 +268,25 @@ pub struct Parameters {
...
 
 impl module::Parameters for Parameters {
-    type Error = std::convert::Infallible;
+    type Error = ParameterValidationError;
+
+    fn validate_basic(&self) -> Result<(), Self::Error> {
+        // Validate dynamic min gas price parameters.
+        let dmgp = &self.dynamic_min_gas_price;
+        if dmgp.enabled {
+            if dmgp.target_block_gas_usage_percentage > 100 {
```
[the remark this answers is not in the document]
Comment (ptrus) [reply runtime-sdk/src/modules/core/mod.rs]: Makes no sense to be zero. I've limited it to (arbitrarily decided) 10-100 range now.  Likely 50 is the most sensible value and could even hardcode it to that (like Ethereum, which has it hardcoded).

File: runtime-sdk/src/modules/core/mod.rs [hunk 5]
```diff
@@ -1040,6 +1146,53 @@ impl<Cfg: Config> module::BlockHandler for Module<Cfg> {
...
+        if !params.dynamic_min_gas_price.enabled {
+            return;
+        }
+
+        // Update dynamic min gas price for next block, inspired by EIP-1559.
+        //
+        // Adjust the min gas price for each block based on the gas used in the previous block and the desired target
+        // gas usage set by `targe_block_gas_usage_percentage`.
+        let gas_used = Self::batch_gas_used(ctx) as u128;
+        let max_batch_gas = Self::max_batch_gas(ctx);
```
[the remark this answers is not in the document]
Comment (ptrus) [reply runtime-sdk/src/modules/core/mod.rs]: Updated :+1:

**Review by kostko — APPROVED** [review]
